Implemented the benchmark to observe the effect of middleware over service-client jobs

CihatAltiparmak opened the pull request:
I couldn't test this scenario with rmw_zenoh but in my measurements, Seems rmw_fastrtps is little bit faster than rmw_cyclonedds. 

Edit: Btw i have realised now that i haven't add documentation.
Edit2: Added documentation

CihatAltiparmak committed:
Implemented the benchmark to observe the effect of middleware over service-client jobs

CihatAltiparmak committed:
Fixed dependency typo for demo_nodes_cpp

CihatAltiparmak committed:
Added the relavant documentation for basic service client benchmarking scenario

CihatAltiparmak committed:
Fix typo in basic_service_client_bechmark documentation

sjahr commented on CMakeLists.txt:
```diff
@@ -28,7 +30,8 @@ ament_target_dependencies(
   "dynmsg"
   "nav_msgs"
-  "yaml-cpp")
+  "yaml-cpp"
```
Nitpick: I don't think you need the `"` here and you can remove it from all CMakeLists.txt files (maybe in a separate small cleanup PR)

CihatAltiparmak replied:
Will be solved in small PR

sjahr commented on docs/how_to_run.md:
```diff
@@ -7,3 +7,9 @@ This benchmark measures the elapsed time by which the determined path is sent fo
+
+This benchmark measures the total elapsed time based on the time interval between sending the request by the client to the server and getting the response of server. This benchmark utilizes the [ros2/demos](https://github.com/ros2/demos) packages' [example server](https://github.com/ros2/demos/blob/rolling/demo_nodes_cpp/src/services/add_two_ints_server.cpp).
+
+In this benchmark scenario, the benchmarker node only has client interface. THe necessary server for this client is run in [the launch file of this benchmark scenario](../launch/scenario_basic_service_client_benchmark.launch.py). Client sends a request to server and waits for the response from server. Client sends second request to server once the client receives response of first request from cl ...
```
```suggestion
In this benchmark scenario, the benchmarker node only has client interface. The necessary server for this client is run in [the launch file of this benchmark scenario](../launch/scenario_basic_service_client_benchmark.launch.py). Client sends a request to server and waits for the response from server. Client sends second request to server once the client receives response of first request from client. This actions are repeated `sending_request_number` times. You can configure this `sending_request_number` parameter in [this scenario's launch file]((../launch/scenario_basic_service_client_benchmark.launch.py)).
```

sjahr commented on docs/scenarios/basic_service_client_benchmark.md:
```diff
@@ -0,0 +1,38 @@
+## How To Run Basic Service Client Benchmark
+
+Firstly, source your ros version. It's suggested to test with rolling version of ROS.
```
```suggestion
Firstly, source your ros version. It's suggested to test with rolling version of ROS 2.
```

sjahr commented on include/moveit_middleware_benchmark/scenarios/scenario_basic_service_client.hpp:
```diff
@@ -0,0 +1,97 @@
+class ScenarioBasicServiceClient
+{
+public:
+  /** \brief Constructor to send request by client to add_two_ints_server which is one
```
Nitpick: If the constructor doesn't do anything unusual and just instantiates members, you can keep it's documentation short by just using

```
/** \brief Constructor
   * \param[in] ...
```

CihatAltiparmak replied:
Done it in this commit ( https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/18/commits/5fdfeaa208a88f57e1e9fe8ff25cc9e6bba47dd3 )

CihatAltiparmak commented:
Now, this PR is on the way @sjahr , could you take a look at it when you're available?

sjahr commented on include/moveit_middleware_benchmark/scenarios/scenario_basic_service_client.hpp:
```diff
@@ -0,0 +1,97 @@
+   * of the example of ros2/demos package
+   *  \param [in] node The ros node for sending request and wait response from server
+   */
+  ScenarioBasicServiceClient(rclcpp::Node::SharedPtr node);
```
Does this make sense?
```suggestion
  ScenarioBasicServiceClient(const rclcpp::Node::SharedPtr& node);
```

CihatAltiparmak replied:
Of course makes sensee. Solved in this commit ( https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/18/commits/f46f58d1344319b1909ed2b4af87f2f74bb88556 )

sjahr reviewed (commented):
Just some minor comments. Tag me or Henning again when this is ready for review

CihatAltiparmak committed:
Fix typos docs/how_to_run.md

Co-authored-by: Sebastian Jahr <sebastian.jahr@tuta.io>

CihatAltiparmak committed:
Fix ROS 2 typo in the doc

Co-authored-by: Sebastian Jahr <sebastian.jahr@tuta.io>

CihatAltiparmak committed:
Kept documentation short for constructor of ScenarioBasicServiceClient class ( see https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/18#discussion_r1702106816 )

CihatAltiparmak committed:
Passed node parameter in const & type. See ( https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/18#discussion_r1702107562 )

CihatAltiparmak commented:
Let's tag in a cool way. @sjahr Would you review again? I'm ready. Btw industrial CI gives failure without any reason. So, i have created this PR to solve this problem. PR is here ( https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/23 ).

CihatAltiparmak committed:
Eexcuted add_two_ints_server in Node instead of ExecuteProcess.

- ExecuteProcess runs two process which are sh and ros2_node,
  if there is no signal handler for SIGINT, SIGTERM etc. when sh is terminated,
  some process may continue to run without killing.

CihatAltiparmak committed:
Added service client benchmark to readme
